### PR TITLE
Implement a Worker Thread Pool

### PR DESCRIPTION
**Pull request by reduz**
This PR implements a thread runner. It uses a fixed amount of threads in a pool and allows scheduling tasks
that can be run on threads (and then waited for). It satisfies the following use cases:

* HTML5 thread count is fixed (and similar restrictions are known in consoles) so we need to reuse threads. 
* Thread spawning is slow in general, so reusing threads is faster anyway.
* This implementation supports recursive waiting for tasks, making it less prone to deadlocks if threads from the pool also run tasks.
* An additional bonus of this is that we can, in the debugger, display how the thread pool is utilized to ensure we are taking advantage of threads as much as possible.

Supported features:
* Adding single tasks or task groups (useful to replace ThreadWorkPool).
* Scheduling high or low priority. On platforms with strong kernel scheduling, low priority tasks use the OS native threads to spawn threads and not exhaust the high priority tasks (as they in most cases will be waiting for IO and suspended by the kernel). On platforms like HTML5, where there is basically no availability of system threads, some of the pool threads will be used for low priority in a configurable % (30% by default) to avoid bottlenecking physics or rendering.

This feature is needed to ensure Godot 4.0 can ship with HTML5 support.

After this is approved and merged, subsequent PRs will be needed to replace the ThreadWorkPool usage by this class.

FAQ:
**Q**: Should interrupting tasks be supported?
**A**: No. Implement this on your own with an an (atomic) bool or something like that if you need it on your own task.
**Q**: Can task be left running while the engine ends?
**A**: No, Godot requires you wait for all tasks to finish before exiting. If you do not, the engine will error and it might crash.

### CODE REVIEW

File: core/object/thread_runner.cpp
Comment (RandomShaper) [inline core/object/thread_runner.cpp:176]: For symmetry with `init()`, maybe this is more appropriate:
```suggestion
	ERR_FAIL_COND(threads.size() == 0);
```

**Review by RandomShaper — CHANGES_REQUESTED** [review]

File: core/object/thread_runner.cpp
Comment (bruvzg) [inline core/object/thread_runner.cpp:148]: I think `Semaphore` is implemented based on `std::mutex` on all platforms, so it need not be platform specific. But probably it's OK as is.

**Review by bruvzg — COMMENTED** [review]
ThreadRunner singleton is definitely a huge improvement over current ThreadWorkPool which starts a thread for each instance and clogging system resources.

File: core/object/thread_runner.cpp
Comment (derammo) [inline core/object/thread_runner.cpp:181]: This isn't really an error.   It should be possible to submit a bunch of long running tasks and then exit the program.  For that to work, you'd need an additional (optional) Callable in add_task(...) so that there is a separate function to call for cancellation.   At shut down, all pending tasks get called on the cancellation callback.  Waiting on a task that was canceled must then also survive without error, because some of the user code may at the same time call wait_for_task_completion(...) during the shutdown process, based on timing.   If ThreadRunner was RefCounted, you could probably just keep all the task objects that were canceled, with the semaphores posted.   Then these deallocate only when all customers of the ThreadRunner are gone.

**Review by derammo — COMMENTED** [review]
I reviewed the code in detail and did not find deadlocks or bugs.  I made a lot of general comments.

File: core/object/thread_runner.h
Comment (derammo) [reply core/object/thread_runner.h:81]: If you check this inside _process_task() and then early return with a bool return value to indicate the thread should exit instead, then this doesn't even need to be atomic.  In the normal case, you check this atomic, and it is false.  Then you call _process_task() which 100% of the time grabs the mutex.   In all cases except on exit, the atomic access isn't useful and is just another synchronization point.

File: core/object/thread_runner.cpp
Comment (derammo) [reply core/object/thread_runner.cpp:55]: Concretely, maybe this object is an internal object only, and the user-visible object is a shim that is assigned to one of N such internal pools.   On a thread restricted platform, all shims would map to the same pool (N=1, with the number of threads set to the platform limit.)  On a large platform, you might allocate several internal pools if you want to always allocate reserved threads to different functional areas.   Not sure you need that, as the OS will likely do a decent job keeping things fair if you just put a lot of threads in the pool on a large platform.

Comment (derammo) [reply core/object/thread_runner.cpp:55]: Game folks might like the idea though of saying "3 threads for this type of task, 7 threads for this type of task."

File: core/object/thread_runner.cpp
Comment (reduz) [reply core/object/thread_runner.cpp:176]: Thats ok, it means if you did not initialize the object, it should not deinitialize it. Some classes are initialized for testing by Godot to get their default values, so its good you can init/deinit safely.

File: core/object/thread_runner.cpp
Comment (reduz) [reply core/object/thread_runner.cpp:97]: I think it should be fine, since the actual one that does polling actually contains wait in the name. This means you could deduce that does not block relatively easily.

File: core/object/thread_runner.cpp
Comment (reduz) [reply core/object/thread_runner.cpp:55]: On a big PC we can just preallocate as many threads as we want, but this API needs to ensure we can work with a small amount for platforms where this is troublesome (HTML5 and some consoles).

File: core/object/thread_runner.cpp
Comment (reduz) [reply core/object/thread_runner.cpp:141]: Yes absolutely, It needs to be proper documented how it works internally.

File: core/object/thread_runner.cpp
Comment (reduz) [reply core/object/thread_runner.cpp:181]: Long running tasks IMO do not belong here and they should use dedicated system threads. I want to work on a specific interface for this separate to this one, since we have some use cases for it. Of course this should be documented properly.

File: core/object/thread_runner.cpp
Comment (derammo) [reply core/object/thread_runner.cpp:97]: Misunderstanding.  I meant the API suggests that client code that wants to poll for status use this function.  As such, they might call this a lot of time, so give them a way to at least not lock for *every* task they are waiting on.  Allow checking all tasks you are waiting for in one go.

File: core/object/thread_runner.cpp
Comment (derammo) [reply core/object/thread_runner.cpp:55]: Yes, so is this a singleton then?  Will it be provided in the namespace by default?

It would be nicer API if users can TaskRunner.new() and just get a shim that calls the single pool that Godot provides.  Don't make the user responsible for only instantiating one of these and have code from different modules/developers share it.

File: core/object/thread_runner.cpp
Comment (derammo) [reply core/object/thread_runner.cpp:148]: I concur, just pointing out Linux eventfd https://man7.org/linux/man-pages/man2/eventfd.2.html was created to let you block on multiple things with poll() or select().  But you really don't want to go there if you have many tasks outstanding potentially.

File: core/object/thread_runner.cpp
Comment (derammo) [reply core/object/thread_runner.cpp:181]: If I read it correctly, all scheduled tasks are in `_tasks_waiting`, even ones that nobody is blocked on.   This could be a lot of stuff, including expensive things like reading files etc.  I think you would want to cancel scheduled work that never started without having to wait for it to actually run?

**Comment by bitsawer** [comment]
There should probably be an API call for cancelling a task (or tasks), too (unless I missed it). Otherwise the creator of those tasks has to stick around and wait for all the tasks he added to run and complete even if he doesn't need them to run anymore (if only to avoid that eror spam in console as @derammo mentioned). Might make node lifetime handling a bit easier when using the runner.

File: core/object/thread_runner.cpp
Comment (derammo) [reply core/object/thread_runner.cpp:181]: More context in case it is useful:

An async API should always have a guarantee of _exactly_ one callback.   This is so that user code can allocate stuff for the async request, stick it in user data, and clean it up on callback.   When the system shuts down, this state needs to be cleaned up, but it may not be possible to execute the queued work, because the system is going down.  

Additionally, it is just slow to have shut down wait for a bunch of work that nobody is going to use anyway.  That's why there normally is a callback parameter that tells you if you are supposed to just clean up or actually do your work, or you can have an additional callback (Callable) to call if the work item / task gets canceled before running.  Because you can't force GD clients to have a particular callback signature, I think the optional Callable is the better way here.  In that case, the guarantee will be that either your work function gets called or the cancellation callback.  Either way, you know you can clean up whatever related stuff you may have.

File: core/object/thread_runner.cpp
Comment (derammo) [inline core/object/thread_runner.cpp:69]: return `INVALID_TASK_ID` if `exit_threads` ?

File: core/object/thread_runner.cpp
Comment (derammo) [inline core/object/thread_runner.cpp:85]: return `INVALID_TASK_ID` if `exit_threads` ?

**Review by derammo — COMMENTED** [review]
Second round after conversation regarding shut down.

**Comment by xix-xeaon** [comment]
> This feature is needed to ensure Godot 4.0 can ship with HTML5 support.

That's great! Which rendering methods and drivers will be supported?

File: core/object/thread_runner.cpp
Comment (reduz) [reply core/object/thread_runner.cpp]: As the lifetime of this object is not controlled by the user, there is no danger of this being called from the wrong place.

Comment (derammo) [reply core/object/thread_runner.cpp]: I was thinking during shut down to catch any late attempts to create more work while we are closing (which would be user code error.)  But if worker threads aren't shut down until the runner is no longer referenced, then this can't happen.

File: core/object/worker_thread_pool.cpp
Comment (jordo) [inline core/object/worker_thread_pool.cpp:313]: Didn't release the lock on error return here.  Will leave `task_mutex` locked if you ask for an invalid group id. (Would make good unit test-case).

File: core/object/worker_thread_pool.cpp
Comment (derammo) [inline core/object/worker_thread_pool.cpp:69]: I assume `arg` was supposed to be set to `work_index`?  Sorry if I missed it.

**Comment by jordo** [comment]
Just a note on int types as well:

```
INVALID_TASK_ID = -1
typedef int64_t TaskID;
typedef int64_t GroupID;
uint64_t last_task = 1;
```

As far as i can tell, `last_task` looks like a global unsigned counter that will wrap,  but then assigned to `TaskID` and `GroupID`  which are signed as the invalid ID was defined as -1.  At some point, potentially say a long running game server, there will be potential for INVALID_TASK_ID to now be valid on wrap around assignment... albeit it would be pretty far out.  

Could define all ID's as unsigned (preferred i think?), define `INVALID_TASK_ID` as `0`, and in the wrap around case, wrap to `1`.

File: core/object/worker_thread_pool.cpp
Comment (reduz) [reply core/object/worker_thread_pool.cpp:69]: ah yes, thanks

File: core/object/worker_thread_pool.cpp
Comment (derammo) [reply core/object/worker_thread_pool.cpp:369]: Makes sense.   It would maybe be worth a comment in the code so future developer doesn't just see the I/O under lock and think it can't possibly be intentional.   In this case, it is clever and a feature, but doesn't read that way when you first see it.  Maybe I am just not smart enough.

**Comment by lyuma** [comment]
I'd be worried about mixing level loading (disk), http (network) and game tasks (cpu) into one "low priority" queue. they are all so dependent on different resources. An example of a slow game tasks which might end up in the low priority queue would be updating some AI navmesh or calculating strategy.

A bottleneck in one resource, such as http, could cause the game to malfunction or appear stuck, even if the http tasks are unimportant, for example submitting metrics to a backend server.

**Comment by reduz** [comment]
@lyuma I am thinking that for low priority or some types of background tasks like network polling, it could be easier to just create a regular OS thread where supported om demand and rely on the scheduler. In platforms where this is not possible, it will use a low priority thread instead.

**Comment by reduz** [comment]
@derammo cancelling IMO is just not possible with this aporoach due to the shared semaphore. Even if the task did not begin processing, the wait will eventually go through because it was posted previously. I am not sure if there is an early way out.

File: core/object/worker_thread_pool.h
Comment (RandomShaper) [inline core/object/worker_thread_pool.h:56]:
```suggestion
		SafeNumeric<uint32_t> index;
```

File: core/object/worker_thread_pool.h
Comment (RandomShaper) [inline core/object/worker_thread_pool.h:60]:
```suggestion
		SafeNumeric<uint32_t> finished;
```

File: core/object/worker_thread_pool.h
Comment (RandomShaper) [inline core/object/worker_thread_pool.h:103]:
```suggestion
	SafeNumeric<uint32_t> low_priority_threads_used;
```

File: tests/core/threads/test_worker_thread_pool.h
Comment (RandomShaper) [inline tests/core/threads/test_worker_thread_pool.h:52]:
```suggestion
static SafeNumeric<uint32_t> callable_counter;
```

File: tests/core/threads/test_worker_thread_pool.h
Comment (RandomShaper) [inline tests/core/threads/test_worker_thread_pool.h:60]:
```suggestion
	SafeNumeric<uint32_t> counter;
```

File: tests/core/threads/test_worker_thread_pool.h
Comment (RandomShaper) [inline tests/core/threads/test_worker_thread_pool.h:92]:
```suggestion
	SafeNumeric<uint32_t> counter;
```

File: tests/core/threads/test_worker_thread_pool.h
Comment (RandomShaper) [inline tests/core/threads/test_worker_thread_pool.h:98]:
```suggestion
static SafeNumeric<uint32_t> callable_group_counter;
```

File: core/object/worker_thread_pool.h
Comment (RandomShaper) [inline core/object/worker_thread_pool.h:59]:
```suggestion
		SafeFlag completed;
```

**Review by RandomShaper — CHANGES_REQUESTED** [review]

File: core/object/worker_thread_pool.cpp
Comment (RandomShaper) [inline core/object/worker_thread_pool.cpp:351]: This is not needed.

File: core/object/worker_thread_pool.h
Comment (RandomShaper) [inline core/object/worker_thread_pool.h:96]: I think other commeter mentioned that this doesn't really need to be that safe. Given that it's synced by the semaphore, we could just have a new `RelaxedFlag` specialization that, while still being atomic, doesn't involve memory ordering constraints. However... we don't have it yet, so I think we can go on and maybe revisit in the future.

**Comment by RandomShaper** [comment]
Besides my nitpicks about the usage of `SafeFlag` and `SafeNumeric`, I don't have anything to say. When this is put into practice we will get more insight about potential improvements or reassessment of the approach to prioritization. I haven't detected any problem in the threading part itself.

File: core/object/worker_thread_pool.h
Comment (RandomShaper) [reply core/object/worker_thread_pool.h:96]: Well, now we can have it if you want to use it: https://github.com/godotengine/godot/pull/63189

**Comment by reduz** [comment]
Okay, I think the implementation should now be final (leaving fixes aside). Once merged, a subsequent PR should be made to replace ThreadWorkPool.

**Review by RandomShaper — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
